examples/rust/components/http-task-manager/src/http.rs

```diff
@@ -0,0 +1,272 @@
+    ($log_ctx:ident, $status_code:expr, $msg:expr, $r_out:expr $(,)?) => {
+        if let Err(e) = crate::http::send_error_response($status_code, $msg, $r_out) {
+            crate::bindings::wasi::logging::logging::log(
+                crate::bindings::wasi::logging::logging::Level::Error,
+                $log_ctx,
+                &format!("failed to send error response: {e}"),
+            );
+            panic!("failed to send error response");
+        }
+    };
+}
+pub(crate) use try_send_error;
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct TaskCompleteRequestBody {
+    pub(crate) worker_id: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct TaskLeaseRequestBody {
+    pub(crate) worker_id: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct TaskReleaseRequestBody {
+    pub(crate) worker_id: String,
+    pub(crate) lease_id: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct TaskFailedRequestBody {
+    pub(crate) worker_id: String,
+    pub(crate) reason: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct TaskSubmitRequestBody {
+    pub(crate) group_id: String,
+    pub(crate) task_data: serde_json::Value,
+}
+
+// NOTE: Since wit-bindgen makes `IncomingRequest` available to us as a local type,
+// we can add convenience functions to it
+impl IncomingRequest {
+    /// Check if the incoming request body has the `application/json` content type
+    pub(crate) fn is_json(&self) -> bool {
+        self.headers()
+            .entries()
+            .iter()
+            .find(|(k, _v)| k.to_lowercase() == "content-type")
+            .is_some_and(|(_k, v)| String::from_utf8_lossy(v).contains("application/json"))
+    }
+
+    /// This is a convenience function that extracts out the body of a IncomingRequest
+    pub(crate) fn read_body(self) -> Result<Bytes> {
+        let incoming_req_body = self
+            .consume()
+            .map_err(|()| anyhow!("failed to consume incoming request body"))?;
+        let incoming_req_body_stream = incoming_req_body
+            .stream()
+            .map_err(|()| anyhow!("failed to build stream for incoming request body"))?;
+        let mut buf = BytesMut::with_capacity(MAX_READ_BYTES);
+        loop {
+            match incoming_req_body_stream.read(MAX_READ_BYTES as u64) {
+                Ok(bytes) if bytes.is_empty() => break,
+                Ok(bytes) => {
+                    ensure!(
+                        bytes.len() <= MAX_READ_BYTES,
+                        "read more bytes than requested"
+                    );
+                    buf.extend(bytes);
+                }
+                Err(StreamError::Closed) => break,
+                Err(e) => bail!("failed to read bytes: {e}"),
+            }
+        }
+        drop(incoming_req_body_stream);
```

Comment:
I think this block could be improved a lot by using the `wasi=0.13.1` crate & just using `io::copy` to copy the stream to the other, but that's fine as a followup!

Reply:
Oh yeah, I absolutely saw the changes ya'll made to do use the `InputStream` stuff earlier I can definitely use it here too